src/components/external-apis/routes.ts

```diff
@@ -0,0 +1,130 @@
+import { FastifyRequest, FastifyReply } from 'fastify'
+import path from 'path'
+import config from '../../config'
+import { get as extractProperty } from 'lodash'
+import { getPermissionNamesFromJWT } from '../data_display/helpers'
+import axios, { AxiosError, AxiosRequestConfig } from 'axios'
+import { constructAuthHeader, constructQueryObject, validateResult } from './helpers'
+import { ExternalApiConfigs, QueryParameters, PostRoute } from './types'
+import { getApplicationData } from '../actions'
+import { getUserInfo } from '../permissions/loginHelpers'
+import { ActionApplicationData } from '../../types'
+import functions from '../actions/evaluatorFunctions'
+import { errorMessage } from '../utilityFunctions'
+
+interface ExternalApiRequest {
+  name: string
+  route: string
+}
+
+const apiConfigs: ExternalApiConfigs = config?.externalApiConfigs ?? {}
+
+export const routeAccessExternalApi = async (
+  request: FastifyRequest & { auth?: { userId?: number; orgId?: number } },
+  reply: FastifyReply
+) => {
+  const { name, route } = request.params as ExternalApiRequest
+
+  const { baseUrl, routes, authentication } = apiConfigs?.[name]
+  if (!baseUrl) {
+    throw new Error('No baseUrl defined for API: ' + name)
+  }
+
+  const routeConfig = routes?.[route]
+  if (!routeConfig) {
+    throw new Error('No config definition for Route: ' + route)
+  }
+
+  const {
+    method,
+    url,
+    permissions,
+    queryParams,
+    allowedClientQueries,
+    returnProperty,
+    additionalAxiosProperties,
+    validationExpression,
+  } = routeConfig
+
+  if (permissions) {
+    const { permissionNames } = await getPermissionNamesFromJWT(request)
+    const validPermissions = permissions.filter((permission) =>
+      permissionNames.includes(permission)
+    )
+
+    if (validPermissions.length === 0) {
+      reply.status(403)
+      return reply.send('Not authorized to view this resource')
+    }
+  }
+
+  // Construct data object for subsequent expression evaluator
+  const { userId, orgId } = request.auth as { userId?: number; orgId?: number }
+  const { user } = await getUserInfo({ userId, orgId })
+  const evaluatorData: {
+    user: typeof user
+    applicationData?: ActionApplicationData
+    functions: typeof functions
+  } = { user, functions }
+
+  // ApplicationData only available if an applicationId is provided as a query
+  // parameter
+  const { applicationId } = request.query as { applicationId?: number }
```

Comment:
can this also be in request type definition in method header ?

Reply:
The "query" property is `unknown` in Fastify types, so we have to define it somewhere, or add a runtime check (which seems unnecessary in this case).

In fact, I'm pretty sure I copied this style of typing the request query with `as` in its first usage from you: 

<img width="673" alt="Screenshot 2023-11-12 at 1 23 25 PM" src="https://github.com/msupply-foundation/conforma-server/assets/5456533/c7002c66-98a0-4e92-9394-6282842a9915">

😉

Reply:
I've been anti ` as ` lately (as i've seen it misused in open mSupply), in general I think ` as ` is fine as long as we don't 'leak' the type further, i.e. when we cast to as and return that type somewhere else and use it, it's hard to validate that the type is indeed correct, have to jump to a lot of places to confirm